doc/user-manual/language/runtime-irrelevance.lagda.rst

```diff
@@ -0,0 +1,139 @@
+
+A function or constructor argument is declared erased using the ``@0`` or ``@erased`` annotation.
+For example, the following definition of vectors guarantees that the length argument to ``_∷_`` is not
+present at runtime::
+
+  data Vec (A : Set a) : Nat → Set a where
+    []  : Vec A 0
+    _∷_ : ∀ {@0 n} → A → Vec A n → Vec A (suc n)
+
+The :ref:`GHC backend <ghc-backend>` compiles this to a datatype where the cons constructor takes only two
+arguments.
+
+.. note::
+  In this particular case, the compiler identifies that the length argument can be erased also without the
+  annotation, using Brady et al's forcing analysis :ref:`[1] <references>`. Marking it erased explictly, however,
+  ensures that it is erased without relying on the analysis.
+
+Erasure annotation can also appear in function arguments (both first-order and higher-order). For instance, here is
+an implementation of ``foldl`` on vectors::
+
+  foldl : (B : Nat → Set b)
+        → (f : ∀ {@0 n} → B n → A → B (suc n))
+        → (z : B 0)
+        → ∀ {@0 n} → Vec A n → B n
+  foldl B f z []       = z
+  foldl B f z (x ∷ xs) = foldl (λ n → B (suc n)) (λ {n} → f {suc n}) (f z x) xs
+
+Here the length arguments to ``foldl`` and to ``f`` have been marked erased. As a result it gets compiled to the following
+Haskell code (modulo renaming):
+
+.. code-block:: text
+
+  foldl f z xs
+    = case xs of
+        []     -> z
+        x ∷ xs -> foldl f (f _ z x) xs
+
+In contrast to constructor arguments, erased arguments to higher-order functions are not removed completely, but
+instead replaced by a placeholder value ``_``. The crucial optimization enabled by the erasure annotation is compiling
+``λ {n} → f {suc n}`` to simply ``f``, removing a terrible space leak from the program. Compare to the result of
+compiling without erasure:
+
+.. code-block:: text
+
+  foldl f z xs
+    = case xs of
+        []     -> z
+        x ∷ xs -> foldl (\ n -> f (1 + n)) (f 0 z x) xs
+
+It is also possible to mark top-level definitions as erased. This guarantees that they are only used in erased arguments
+and can be useful to ensure that code intended only for compile-time evaluation is not executed at run time. For instance,
+
+::
+
+  @0 spec : Nat → Nat   -- slow, but easy to verify
+  impl    : Nat → Nat   -- fast, but hard to understand
+  proof   : ∀ n → spec n ≡ impl n
+
+..
+  ::
+  spec n = n
+  impl n = n
+  proof n = refl
+
+Rules
+=====
+
+The typing rules are based on Conor McBride's "I Got Plenty o’Nuttin’" :ref:`[2] <references>`. In essence the type checker
+keeps track of whether it is running in *run-time mode*, checking something that is needed at run time, or *compile-time mode*,
+checking something that will be erased. In compile-time mode everything to do with erasure can safely be ignored, but in run-time
+mode the following restrictions apply:
+
+- Cannot use erased variables or defintions
+- Cannot pattern match on erased arguments, unless there is at most one valid case
```

Comment:
This is when the K rule is on. Otherwise, when there is exactly one valid case, pattern matching is allowed if the type is not an inductive family, to rule out pattern matching on erased equality proofs (`refl`) in a setting with univalence.